### DISCUSSION TITLE
Question about local variable limitations with `Binding#eval` in mruby

### DISCUSSION THREAD

**Discussion by buty4649** [mruby/mruby#6274] [Q&A] [answered]
When executing the following code using `Binding#eval`, I encounter the error `-e:1:in eval: too many local variables for binding (mruby limitation) (RuntimeError)`.

```ruby
$ ./bin/mruby -e 'b=binding; 1000.times{ b.eval("i||=0;i+=1"); }'
trace (most recent call last):
        [2] -e:1
        [1] -e:1
-e:1:in eval: too many local variables for binding (mruby limitation) (RuntimeError)
```

Even if I declare the variable `i` only once, the same error occurs.

```ruby
$ ./bin/mruby -e 'b=binding; b.eval("i||=0"); 1000.times{ b.eval("i+=1"); }'
trace (most recent call last):
        [2] -e:1
        [1] -e:1
-e:1:in eval: too many local variables for binding (mruby limitation) (RuntimeError)
```

I understand from the error message that this is a limitation of mruby. Additionally, I am aware that increasing `MRB_IREP_LVAR_MERGE_LIMIT` can increase the number of definitions, but `IREP_LVAR_MERGE_MAXIMUM` (240) is the upper limit.

Since this error is problematic for my use case, I would like to know the following:

1. Why is `IREP_LVAR_MERGE_MAXIMUM` set to a maximum of 240? Is this due to memory constraints?
2. Why does declaring the same variable name multiple times result in a `too many local variables` error?

Any insights or solutions would be greatly appreciated.

---
以下のコードを`Binding#eval`を使って実行すると、`-e:1:in eval: too many local variables for binding (mruby limitation) (RuntimeError)`というエラーが発生します。

```ruby
$ ./bin/mruby -e 'b=binding; 1000.times{ b.eval("i||=0;i+=1"); }'
trace (most recent call last):
        [2] -e:1
        [1] -e:1
-e:1:in eval: too many local variables for binding (mruby limitation) (RuntimeError)
```

変数`i`の宣言を一回にしても同じエラーが発生します。

```ruby
$ ./bin/mruby -e 'b=binding; b.eval("i||=0"); 1000.times{ b.eval("i+=1"); }'
trace (most recent call last):
        [2] -e:1
        [1] -e:1
-e:1:in eval: too many local variables for binding (mruby limitation) (RuntimeError)
```

このエラーメッセージからもわかるように、これはmrubyの制限によるものです。さらに、`MRB_IREP_LVAR_MERGE_LIMIT`を増やせば定義数を増やせることは知っていますが、`IREP_LVAR_MERGE_MAXIMUM`（240）が上限となっています。

私のユースケースでは、このエラーが発生すると困ってしまうため、以下の点についてお聞きしたいです。

1. `IREP_LVAR_MERGE_MAXIMUM`が240という上限になっているのはなぜでしょうか？これはメモリの制限によるものでしょうか？
2. 同じ変数名でも複数回宣言すると`too many local variables`エラーが発生してしまうのはなぜでしょうか？

どなたかご存知の方がいらっしゃいましたら、ご教示いただけると助かります。

**Comment by dearblue** [comment, marked as answer]
>    1. Why is `IREP_LVAR_MERGE_MAXIMUM` set to a maximum of 240? Is this due to memory constraints?

The `mrbgems/mruby-compiler` itself has its own limits, so I set the limits a little stricter than that. If I actually try to define 255 variables, I get an error. The reason for the separation into `v` and `w` in this example is that they are caught by another limit.

`mrbgems/mruby-compiler` 自体にも制限があるので、それよりも少し厳しい制限値にしました。実際に 255 個の変数を定義しようとすると、エラーが発生します。この例で `v` と `w` に分けている理由は、別の制限に引っかかるためです。

```console
% bin/mruby -e 'eval 250.times.with_object("") { |i, a| a << "v#{i}=" } << "12345; w1=w2=w3=w4=98765"; p "OK!"'
"OK!"

% bin/mruby -e 'eval 250.times.with_object("") { |i, a| a << "v#{i}=" } << "12345; w1=w2=w3=w4=w5=98765"; p "OK!"'
(eval):1: too many local variables
trace (most recent call last):
        [1] -e:1
-e:1:in eval: codegen error (ScriptError)
```

>    2. Why does declaring the same variable name multiple times result in a `too many local variables` error?

This is a bug and will be fixed.
すいません、僕がやりました。

**Reply by buty4649** [reply]
返信ありがとうございます！ `mrbgems/mruby-compiler` の制限で255個の変数が上限であることを理解しました。

またこの投稿でバグが発覚してよかったです。
早速修正のPRもありがとうございます。
私のユースケースであれば、#6275がマージされれば解決されそうです 😄

**Comment by matz** [comment]
#6275 merged.
